### PR TITLE
Stop loading m4a files with unsupported sample block sizes

### PR DESCRIPTION
**Pull request by nealey**
I have a whole lot of HE-AAC files with 2048 frames per sample block. They usually play all right if you start at the beginning, but if you seek, all sorts of errors start spitting out, and there is no playback. This patch prevents Mixxx from even loading these files.

I'll try to see if I can work up a second patch which adds support for HE-AAC.

### CODE REVIEW

File: plugins/soundsourcem4a/soundsourcem4a.cpp [hunk 1]
```diff
@@ -126,6 +126,11 @@ Result SoundSourceM4A::tryOpen(const AudioSourceConfig& audioSrcCfg) {
         return ERR;
     }
 
+    if (MP4GetTrackFixedSampleDuration(m_hFile, m_trackId) != kFramesPerSampleBlock) {
```
Comment (ywwg) [inline plugins/soundsourcem4a/soundsourcem4a.cpp]: put a TODO with a link to a launchpad bug for adding support for this

Comment (nealey) [inline plugins/soundsourcem4a/soundsourcem4a.cpp]: Not sure what you mean by "put a TODO" but I just created a Launchpad bug: https://bugs.launchpad.net/mixxx/+bug/1473690

Comment (ywwg) [inline plugins/soundsourcem4a/soundsourcem4a.cpp]: add a comment:

// TODO(lp:#XXXXX): Remove this error condition when we support this type of file

**Comment by nealey** [comment]
Here's a version that actually fixes the problem.

(Sorry, I didn't realize a second commit to this branch would be automatically added to the pull request. I'd intended to submit a second pull request for this.)

**Comment by uklotzde** [comment]
The comment on kNumberOfPrefetchSampleBlocks still refers to a block size of 1024 samples for the calculation. Maybe we should mention that. The deduction of the value remains valid and does not depend on the actual block size.

Thank you for this fix, Neale! Good finding. I don't know why I didn't replace the hard-coded constant with the actual value in the first place.

File: plugins/soundsourcem4a/soundsourcem4a.cpp [hunk 2]
```diff
@@ -126,6 +110,12 @@ Result SoundSourceM4A::tryOpen(const AudioSourceConfig& audioSrcCfg) {
         return ERR;
     }
 
+    m_framesPerSampleBlock = MP4GetTrackFixedSampleDuration(m_hFile, m_trackId);
+    if (MP4_INVALID_DURATION == m_framesPerSampleBlock) {
+      qWarning() << "No support for variable duration track samples";
```
Comment (daschuer) [inline plugins/soundsourcem4a/soundsourcem4a.cpp]: The warning does somehow not match the if condition. 
Can you add a comment to the code. Why this warning is issued here?

**Comment by daschuer** [comment]
Thank you for the fix. 

In order to commit the this PR. We need you permission. 
Please sign: https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
And comment here when done.

**Comment by nealey** [comment]
We may actually need to move the calculation of kNumberOfPrefetchSampleBlocks into code. According to my research, AAC-LD uses 480 frames per sample block. Decoding 3 blocks of 480 frames will only get 1440 samples.

I'd be happy to add this to my patch.

**Comment by uklotzde** [comment]
I read the linked article again and indeed Apple recommends "...2112 samples 
of priming..." independent of the block size. A block size of 1024 is just 
mentioned for the example, because it is the one most commonly used.

Yes, looks like the the number of prefetch block is not constant anymore:

const SINT kNumberOfPrefetchFrames = 2112;

m_numberOfPrefetchSampleBlocks = (kNumberOfPrefetchFrames + 
(m_framesPerSampleBlock - 1)) / m_framesPerSampleBlock;

**Comment by daschuer** [comment]
ok to test

**Comment by daschuer** [comment]
You have lost me ... Can you add a reference to the source docu of 2112? 
@uklotzde: is it OK now?

**Comment by uklotzde** [comment]
I just found 2 minor cosmetic issues: Reordering of members and 
initialization in initializer list.

**Comment by uklotzde** [comment]
LGTM

@daschuer: The URL reference can still be found in the comments about kNumberOfPrefetchFrames

**Comment by daschuer** [comment]
Thank you very much @nealey and @uklotzde

**Pull request merged**